### PR TITLE
Reset a test site

### PR DESCRIPTION
**Pull request by rrennick**
### Changes proposed in this Pull Request:

The goal of this PR is to incorporate resetting the test site into the core E2E tests. It makes the following changes

- introduces a utility flow `withRestApi`
- moves `deleteAllCoupons` and `deleteAllProducts` from the local jest setup to `withRestApi` to allow other projects to take advantage of those functions
- adds an `expect().toBeInRange()` matcher to allow setting up shipping after a reset. See https://github.com/woocommerce/woocommerce-admin/issues/6761
- fixes an issue where the city is duplicated on retest `San FranciscoSan Francisco`
- changes `runSetupWizard()` to not rely on redirect when testing in a reset site

### How to test the changes in this Pull Request:

1. Run `npm install`
2. Run `npx wc-e2e docker:up`
3. Run `npx wc-e2e test:e2e`
4. Tests should pass
5. Run `E2E_RETEST=1 npx wc-e2e test:e2e`
6. Tests should pass 
7. Review changes and additions to package readme and change log files

### Changelog entry

N/A

### CODE REVIEW

File: tests/e2e/utils/src/flows/constants.js [hunk 1]
```diff
@@ -2,32 +2,60 @@
...
-export const WP_ADMIN_ANALYTICS_PAGES = baseUrl + 'wp-admin/admin.php?page=wc-admin&path=%2Fanalytics%2F';
-export const WP_ADMIN_ALL_USERS_VIEW = baseUrl + 'wp-admin/users.php';
-export const WP_ADMIN_IMPORT_PRODUCTS = baseUrl + 'wp-admin/edit.php?post_type=product&page=product_importer';
-
+export const WP_ADMIN_DASHBOARD = baseUrl + 'wp-admin/';
+export const WP_ADMIN_PLUGINS = WP_ADMIN_DASHBOARD + 'plugins.php';
+export const WP_ADMIN_PERMALINK_SETTINGS = WP_ADMIN_DASHBOARD + 'options-permalink.php';
+export const WP_ADMIN_ALL_USERS_VIEW = WP_ADMIN_DASHBOARD + 'users.php';
+/**
+ * WooCommerce core post type pages.
```
Comment (zhongruige) [inline tests/e2e/utils/src/flows/constants.js:17]: I like that you broke these out into logical sections--thanks for making these change!

File: tests/e2e/utils/src/flows/constants.js [hunk 2]
```diff
@@ -2,32 +2,60 @@
...
+export const MY_ACCOUNT_DOWNLOADS = SHOP_MY_ACCOUNT_PAGE + 'downloads';
+export const MY_ACCOUNT_ADDRESSES = SHOP_MY_ACCOUNT_PAGE + 'edit-address';
+export const MY_ACCOUNT_ACCOUNT_DETAILS = SHOP_MY_ACCOUNT_PAGE + 'edit-account';
 
-export const MY_ACCOUNT_ORDERS = baseUrl + 'my-account/orders';
-export const MY_ACCOUNT_DOWNLOADS = baseUrl + 'my-account/downloads';
-export const MY_ACCOUNT_ADDRESSES = baseUrl + 'my-account/edit-address';
-export const MY_ACCOUNT_ACCOUNT_DETAILS = baseUrl + 'my-account/edit-account';
+/**
+ * Test control flags.
```
Comment (zhongruige) [inline tests/e2e/utils/src/flows/constants.js:58]: Do you think in the future we'll want to include the `SMOKE_TEST_*` variables in here?

**Review by zhongruige — APPROVED** [review]
This is awesome @rrennick! Thanks for working on this. Just wanted to note the tests passed for me locally, including the re-run where things were set up again:

Screen Shot 2021-04-14 at 2 13 46 PM

I got curious and wanted to mentioned that the tests also ran a bit faster. I ran ~5 iterations of the tests and we're saving roughly 20-30s. Here's an example run:

Before this branch:

Screen Shot 2021-04-14 at 3 22 30 PM

On this branch:

Screen Shot 2021-04-14 at 2 10 54 PM

File: tests/e2e/utils/src/flows/constants.js (hunk 2 above)
Comment (rrennick) [reply tests/e2e/utils/src/flows/constants.js:58]: > Do you think in the future we'll want to include the SMOKE_TEST_* variables in here?

The smoke test variables all replace counterparts in `default.json`. I'm not sure there is a need to be able to determine whether the config value is the env variable or from the config file.

**Pull request merged**